File: solar_mqtt/solar/scraper.py
"""EZ-Solar 登入 + API 擷取。每個廠一個 Scraper 實例。"""

from __future__ import annotations

from typing import Any

import requests
from bs4 import BeautifulSoup


def _to_float(v: Any) -> float | None:
    try:
        return float(v)
    except (ValueError, TypeError):
        return None


class Scraper:
    """單廠擷取器。封裝 session + zone serial 映射。"""

    def __init__(
        self,
        factory_id: str,
        base_url: str,
        login_user: str,
        login_pass: str,
    ) -> None:
        self.factory_id = factory_id
        self.base_url = base_url.rstrip("/")
        self.login_user = login_user
        self.login_pass = login_pass
        self.session: requests.Session | None = None
        self._zone_serial_map: dict[str, int] = {}
        self._next_zone_id: int = 1

    # ── 更新登入資訊（/set 變更後呼叫） ──
    def update_credentials(
        self,
        base_url: str,
        login_user: str,
        login_pass: str,
    ) -> None:
        self.base_url = base_url.rstrip("/")
        self.login_user = login_user
        self.login_pass = login_pass
        self.session = None  # 強制下一輪重新登入

    # ── 穩定 zone id ──
    def _stable_zone_id(self, serial: str, fallback_idx: int) -> int:
        key = (serial or "").strip() or f"__pos_{fallback_idx}"
        if key not in self._zone_serial_map:
            self._zone_serial_map[key] = self._next_zone_id
            self._next_zone_id += 1
        return self._zone_serial_map[key]

    # ── 登入 ──
    def login(self) -> requests.Session:
        login_url = f"{self.base_url}/default.aspx"
        print(f"[{self.factory_id}] 登入中... {self.base_url}")

        s = requests.Session()
        s.headers.update(
            {
                "User-Agent": (
                    "Mozilla/5.0 (Windows NT 10.0; Win64; x64) "
                    "AppleWebKit/537.36 (KHTML, like Gecko) "
                    "Chrome/120.0 Safari/537.36"
                ),
                "Accept": "text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8",
                "Accept-Language": "zh-TW,zh;q=0.9,en;q=0.8",
            }
        )

        r = s.get(login_url, timeout=10)
        r.raise_for_status()
        soup = BeautifulSoup(r.text, "html.parser")

        form = soup.find("form") or soup
        form_fields: list[tuple[str, str]] = []
        for tag in form.find_all("input"):
            name = tag.get("name")
            if not name:
                continue
            itype = (tag.get("type") or "text").lower()
            if itype == "submit" and name != "defbtn1":
                continue
            value = tag.get("value", "")
            if name == "deftxt1":
                value = self.login_user
            elif name == "deftxt2":
                value = self.login_pass
            form_fields.append((name, value))

        names = {n for n, _ in form_fields}
        if "deftxt1" not in names:
            form_fields.append(("deftxt1", self.login_user))
        if "deftxt2" not in names:
            form_fields.append(("deftxt2", self.login_pass))
        if "defbtn1" not in names:
            form_fields.append(("defbtn1", "登入"))

        headers = {
            "Referer": login_url,
            "Origin": self.base_url,
            "Content-Type": "application/x-www-form-urlencoded; charset=UTF-8",
        }
        r = s.post(login_url, data=form_fields, headers=headers, timeout=10, allow_redirects=True)

        if r.status_code >= 400:
            self._dump_debug(r, f"debug_login_fail_{self.factory_id}.html")
            r.raise_for_status()

        if 'id="deftxt1"' in r.text or 'name="deftxt1"' in r.text:
            self._dump_debug(r, f"debug_login_fail_{self.factory_id}.html")
            raise RuntimeError(f"[{self.factory_id}] 登入失敗：仍停留在登入頁")

        print(f"[{self.factory_id}] 登入完成（最終 URL：{r.url}）")
        self.session = s
        return s

    @staticmethod
    def _dump_debug(r: requests.Response, path: str) -> None:
        try:
            with open(path, "w", encoding="utf-8") as f:
                f.write(f"<!-- status={r.status_code} url={r.url} -->\n")
                f.write(r.text)
            print(f"已寫出 {path}")
        except Exception:
            pass

    # ── API ──
    def _api_post(self, payload_id: str) -> list[dict]:
        assert self.session is not None
        url = f"{self.base_url}/api/s_json.ashx"
        r = self.session.post(url, data={"id": payload_id}, timeout=10)
        r.raise_for_status()
        data = r.json()
        if not isinstance(data, list):
            raise ValueError(f"API {payload_id} 回傳非陣列: {type(data).__name__}")
        return data

    def fetch_summary(self) -> dict:
        data = self._api_post("00_00")
        if not data:
            return {"total_power_kw": None, "today_mwh": None, "month_mwh": None}
        d = data[0]
        return {
            "total_power_kw": _to_float(d.get("x0")),
            "today_mwh":      _to_float(d.get("x1")),
            "month_mwh":      _to_float(d.get("x2")),
        }

    def fetch_zones(self) -> list[dict]:
        data = self._api_post("00_02")
        zones: list[dict] = []
        for idx, d in enumerate(data, start=1):
            today_kwh = _to_float(d.get("x5"))
            capacity_kwp = _to_float(d.get("x12"))
            today_hours = None
            if today_kwh is not None and capacity_kwp and capacity_kwp > 0:
                today_hours = round(today_kwh / capacity_kwp, 2)
            serial = str(d.get("x8") or "").strip()
            zones.append(
                {
                    "zone_id":      self._stable_zone_id(serial, idx),
                    "serial":       serial,
                    "position":     idx,
                    "name":         (d.get("x0") or "").strip(),
                    "power_kw":     _to_float(d.get("x4")),
                    "today_kwh":    today_kwh,
                    "month_mwh":    _to_float(d.get("x6")),
                    "total_mwh":    _to_float(d.get("x7")),
                    "capacity_kwp": capacity_kwp,
                    "today_hours":  today_hours,
                }
            )
        return zones

    # ── 一次性擷取（自動登入） ──
    def fetch(self) -> tuple[dict, list[dict]]:
        if self.session is None:
            self.login()
        return self.fetch_summary(), self.fetch_zones()

File: solar_mqtt_go/internal/scraper/scraper.go
package scraper

import (
	"bytes"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"math"
	"net/http"
	"net/http/cookiejar"
	"net/url"
	"os"
	"strconv"
	"strings"
	"time"

	"golang.org/x/net/html"
)

// Summary 對應 Python fetch_summary() 回傳 dict。
// TotalMwh 由 service 層的完整性 guard 計算後填入。
type Summary struct {
	TotalPowerKw *float64 `json:"total_power_kw"`
	TodayMwh     *float64 `json:"today_mwh"`
	MonthMwh     *float64 `json:"month_mwh"`
	TotalMwh     *float64 `json:"total_mwh,omitempty"`
}

// Zone 對應 Python fetch_zones() 的每區 dict。
type Zone struct {
	ZoneID      int      `json:"zone_id"`
	Serial      string   `json:"serial"`
	Position    int      `json:"position"`
	Name        string   `json:"name"`
	PowerKw     *float64 `json:"power_kw"`
	TodayKwh    *float64 `json:"today_kwh"`
	MonthMwh    *float64 `json:"month_mwh"`
	TotalMwh    *float64 `json:"total_mwh"`
	CapacityKwp *float64 `json:"capacity_kwp"`
	TodayHours  *float64 `json:"today_hours"`
}

// FloatPtr 取值輔助。
func FloatPtr(v float64) *float64 { return &v }

// toFloatPtr 對應 Python _to_float：可轉數字回指標，否則 nil。
// JSON 數字 → float64；字串 → ParseFloat；bool → Python float(True/False)；其他 → nil。
func toFloatPtr(v any) *float64 {
	switch x := v.(type) {
	case nil:
		return nil
	case float64:
		return &x
	case string:
		f, err := strconv.ParseFloat(strings.TrimSpace(x), 64)
		if err != nil {
			return nil
		}
		return &f
	case bool:
		if x {
			f := 1.0
			return &f
		}
		f := 0.0
		return &f
	default:
		return nil
	}
}

// TodayHours 計算今日有效時數：kwh/kwp 四捨五入 2 位；kwp 缺失或不為正 → nil。
// 以 IEEE 格式化達成與 Python round(x, 2) 相同的 half-even 行為。
func TodayHours(todayKwh, capacityKwp float64) *float64 {
	if todayKwh == 0 && capacityKwp <= 0 {
		return nil
	}
	if capacityKwp <= 0 {
		return nil
	}
	s := strconv.FormatFloat(todayKwh/capacityKwp, 'f', 2, 64)
	f, err := strconv.ParseFloat(s, 64)
	if err != nil {
		return nil
	}
	return &f
}

// ── 穩定 zone id ──

func (s *Scraper) stableZoneID(serial string, fallbackIdx int) int {
	key := strings.TrimSpace(serial)
	if key == "" {
		key = fmt.Sprintf("__pos_%d", fallbackIdx)
	}
	if id, ok := s.zoneSerialMap[key]; ok {
		return id
	}
	id := s.nextZoneID
	s.zoneSerialMap[key] = id
	s.nextZoneID++
	return id
}

// ── Scraper ──

// Scraper 單廠擷取器：session + zone serial 映射。
type Scraper struct {
	FactoryID string
	BaseURL   string
	LoginUser string
	LoginPass string

	client *http.Client

	zoneSerialMap map[string]int
	nextZoneID    int
}

// New 建立未登入的 Scraper。
func New(factoryID, baseURL, loginUser, loginPass string) *Scraper {
	return &Scraper{
		FactoryID:     factoryID,
		BaseURL:       strings.TrimRight(baseURL, "/"),
		LoginUser:     loginUser,
		LoginPass:     loginPass,
		zoneSerialMap: map[string]int{},
		nextZoneID:    1,
	}
}

// LoggedIn 回傳是否已有可用 session。
func (s *Scraper) LoggedIn() bool { return s.client != nil }

// ResetSession 清掉現有 session，下一輪強制重新登入（抓取失敗後呼叫）。
func (s *Scraper) ResetSession() { s.client = nil }

// UpdateCredentials 更新登入資訊並強制下一輪重新登入（對應 Python /set 變更）。
func (s *Scraper) UpdateCredentials(baseURL, loginUser, loginPass string) {
	s.BaseURL = strings.TrimRight(baseURL, "/")
	s.LoginUser = loginUser
	s.LoginPass = loginPass
	s.client = nil
}

// setBrowserHeaders 對應 Python session headers。
func setBrowserHeaders(req *http.Request) {
	req.Header.Set("User-Agent",
		"Mozilla/5.0 (Windows NT 10.0; Win64; x64) "+
			"AppleWebKit/537.36 (KHTML, like Gecko) "+
			"Chrome/120.0 Safari/537.36")
	req.Header.Set("Accept", "text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8")
	req.Header.Set("Accept-Language", "zh-TW,zh;q=0.9,en;q=0.8")
}

// dumpDebug 寫出 debug_login_fail_{factory_id}.html（Python 相同格式）。
func (s *Scraper) dumpDebug(status int, finalURL, body string) {
	path := fmt.Sprintf("debug_login_fail_%s.html", s.FactoryID)
	var buf bytes.Buffer
	fmt.Fprintf(&buf, "<!-- status=%d url=%s -->\n", status, redactURL(finalURL))
	buf.WriteString(body)
	if err := os.WriteFile(path, buf.Bytes(), 0o644); err == nil {
		fmt.Printf("已寫出 %s\n", path)
	}
}

func redactURL(raw string) string {
	parsed, err := url.Parse(raw)
	if err != nil {
		return "<invalid URL>"
	}
	parsed.User = nil
	return parsed.String()
}

func redactErrorText(rawURL string, err error) string {
	if err == nil {
		return ""
	}
	return strings.ReplaceAll(err.Error(), rawURL, redactURL(rawURL))
}

func requestError(action, rawURL string, err error) error {
	return fmt.Errorf("%s（%s）: %s", action, redactURL(rawURL), redactErrorText(rawURL, err))
}

// readBody 讀取回應 body（上限 10MB 以防異常回應）。
func readBody(r *http.Response) (string, error) {
	defer r.Body.Close()
	data, err := io.ReadAll(io.LimitReader(r.Body, 10<<20))
	if err != nil {
		return "", err
	}
	return string(data), nil
}

// newSession 建立帶 cookie jar 的 client（10 秒 timeout，對應 Python timeout=10）。
func newSession() (*http.Client, error) {
	jar, err := cookiejar.New(nil)
	if err != nil {
		return nil, err
	}
	return &http.Client{
		Timeout: 10 * time.Second,
		Jar:     jar,
		CheckRedirect: func(req *http.Request, via []*http.Request) error {
			if len(via) >= 10 {
				return errors.New("too many redirects")
			}
			return nil
		},
	}, nil
}

// collectFormFields 解析登入頁：找第一個 form（無則整份文件），
// 收集 input 欄位；submit 型態僅保留 defbtn1；deftxt1/deftxt2 填入帳密。
func collectFormFields(body string, loginUser, loginPass string) []string {
	var fields []string
	seen := map[string]bool{}

	add := func(name, value string) {
		fields = append(fields, url.QueryEscape(name)+"="+url.QueryEscape(value))
		seen[name] = true
	}

	doc, err := html.Parse(strings.NewReader(body))
	if err != nil {
		// 無法解析 → 走「補齊三欄位」路徑
		add("deftxt1", loginUser)
		add("deftxt2", loginPass)
		add("defbtn1", "登入")
		return fields
	}

	var form *html.Node
	var findForm func(n *html.Node)
	findForm = func(n *html.Node) {
		if form != nil {
			return
		}
		if n.Type == html.ElementNode && n.Data == "form" {
			form = n
			return
		}
		for c := n.FirstChild; c != nil; c = c.NextSibling {
			findForm(c)
			if form != nil {
				return
			}
		}
	}
	findForm(doc)

	root := form
	if root == nil {
		root = doc
	}

	var walk func(n *html.Node)
	walk = func(n *html.Node) {
		if n.Type == html.ElementNode && n.Data == "input" {
			var name, itype, value string
			for _, a := range n.Attr {
				switch a.Key {
				case "name":
					name = a.Val
				case "type":
					itype = a.Val
				case "value":
					value = a.Val
				}
			}
			if name == "" {
				return
			}
			if itype == "" {
				itype = "text"
			}
			itype = strings.ToLower(itype)
			if itype == "submit" && name != "defbtn1" {
				return
			}
			switch name {
			case "deftxt1":
				value = loginUser
			case "deftxt2":
				value = loginPass
			}
			add(name, value)
		}
		for c := n.FirstChild; c != nil; c = c.NextSibling {
			walk(c)
		}
	}
	walk(root)

	if !seen["deftxt1"] {
		add("deftxt1", loginUser)
	}
	if !seen["deftxt2"] {
		add("deftxt2", loginPass)
	}
	if !seen["defbtn1"] {
		add("defbtn1", "登入")
	}
	return fields
}

// Login 登入 EZ-Solar 並建立 session（對應 Python login()）。
func (s *Scraper) Login() error {
	loginURL := s.BaseURL + "/default.aspx"
	fmt.Printf("[%s] 登入中... %s\n", s.FactoryID, redactURL(s.BaseURL))

	client, err := newSession()
	if err != nil {
		return err
	}

	// GET 登入頁
	req, err := http.NewRequest(http.MethodGet, loginURL, nil)
	if err != nil {
		return requestError("建立登入頁 request 失敗", loginURL, err)
	}
	setBrowserHeaders(req)
	resp, err := client.Do(req)
	if err != nil {
		return requestError("取得登入頁失敗", loginURL, err)
	}
	body, err := readBody(resp)
	if err != nil {
		return err
	}
	if resp.StatusCode >= 400 {
		return fmt.Errorf("取得登入頁失敗: status %d", resp.StatusCode)
	}

	fields := collectFormFields(body, s.LoginUser, s.LoginPass)
	formData := strings.Join(fields, "&")

	// POST 登入
	postReq, err := http.NewRequest(http.MethodPost, loginURL, strings.NewReader(formData))
	if err != nil {
		return requestError("建立登入 POST request 失敗", loginURL, err)
	}
	setBrowserHeaders(postReq)
	postReq.Header.Set("Referer", loginURL)
	postReq.Header.Set("Origin", s.BaseURL)
	postReq.Header.Set("Content-Type", "application/x-www-form-urlencoded; charset=UTF-8")

	postResp, err := client.Do(postReq)
	if err != nil {
		return requestError("登入 POST 失敗", loginURL, err)
	}
	postBody, err := readBody(postResp)
	if err != nil {
		return err
	}

	if postResp.StatusCode >= 400 {
		s.dumpDebug(postResp.StatusCode, redactURL(postResp.Request.URL.String()), postBody)
		return fmt.Errorf("[%s] 登入失敗: status %d", s.FactoryID, postResp.StatusCode)
	}

	if strings.Contains(postBody, `id="deftxt1"`) || strings.Contains(postBody, `name="deftxt1"`) {
		s.dumpDebug(postResp.StatusCode, redactURL(postResp.Request.URL.String()), postBody)
		return fmt.Errorf("[%s] 登入失敗：仍停留在登入頁", s.FactoryID)
	}

	fmt.Printf("[%s] 登入完成（最終 URL：%s）\n", s.FactoryID, redactURL(postResp.Request.URL.String()))
	s.client = client
	return nil
}

// apiPost 對應 Python _api_post：POST api/s_json.ashx 並要求回傳 JSON 陣列。
func (s *Scraper) apiPost(payloadID string) ([]any, error) {
	if s.client == nil {
		return nil, errors.New("尚未登入（session 不存在）")
	}
	apiURL := s.BaseURL + "/api/s_json.ashx"
	form := url.Values{}
	form.Set("id", payloadID)

	req, err := http.NewRequest(http.MethodPost, apiURL, strings.NewReader(form.Encode()))
	if err != nil {
		return nil, requestError("建立 API request 失敗", apiURL, err)
	}
	req.Header.Set("Content-Type", "application/x-www-form-urlencoded")

	resp, err := s.client.Do(req)
	if err != nil {
		return nil, requestError("API 請求失敗", apiURL, err)
	}
	body, err := readBody(resp)
	if err != nil {
		return nil, err
	}
	if resp.StatusCode >= 400 {
		return nil, fmt.Errorf("API %s 失敗: status %d", payloadID, resp.StatusCode)
	}
	var data []any
	if err := json.Unmarshal([]byte(body), &data); err != nil {
		return nil, fmt.Errorf("API %s 回傳非 JSON 陣列: %v", payloadID, err)
	}
	return data, nil
}

// FetchSummary 對應 Python fetch_summary()。
func (s *Scraper) FetchSummary() (*Summary, error) {
	data, err := s.apiPost("00_00")
	if err != nil {
		return nil, err
	}
	sm := &Summary{}
	if len(data) == 0 {
		return sm, nil
	}
	d, ok := data[0].(map[string]any)
	if !ok {
		return sm, nil
	}
	sm.TotalPowerKw = toFloatPtr(d["x0"])
	sm.TodayMwh = toFloatPtr(d["x1"])
	sm.MonthMwh = toFloatPtr(d["x2"])
	return sm, nil
}

// strField 對應 Python str(d.get(k) or "").strip()：falsy（nil、0、false、""）
// → 空字串；其他純量 → str() 後 trim。
func strField(v any) string {
	switch x := v.(type) {
	case nil:
		return ""
	case string:
		if x == "" {
			return ""
		}
		return strings.TrimSpace(x)
	case float64:
		if x == 0 {
			return ""
		}
		return strings.TrimSpace(strconv.FormatFloat(x, 'f', -1, 64))
	case bool:
		if !x {
			return ""
		}
		return "True"
	default:
		return strings.TrimSpace(fmt.Sprint(v))
	}
}

// FetchZones 對應 Python fetch_zones()。
func (s *Scraper) FetchZones() ([]Zone, error) {
	data, err := s.apiPost("00_02")
	if err != nil {
		return nil, err
	}
	zones := make([]Zone, 0, len(data))
	for idx, item := range data {
		d, ok := item.(map[string]any)
		if !ok {
			continue
		}
		position := idx + 1
		serial := strField(d["x8"])
		name := strField(d["x0"])
		z := Zone{
			ZoneID:      s.stableZoneID(serial, position),
			Serial:      serial,
			Position:    position,
			Name:        strings.TrimSpace(name),
			PowerKw:     toFloatPtr(d["x4"]),
			TodayKwh:    toFloatPtr(d["x5"]),
			MonthMwh:    toFloatPtr(d["x6"]),
			TotalMwh:    toFloatPtr(d["x7"]),
			CapacityKwp: toFloatPtr(d["x12"]),
		}
		if z.TodayKwh != nil && z.CapacityKwp != nil && *z.CapacityKwp > 0 {
			z.TodayHours = TodayHours(*z.TodayKwh, *z.CapacityKwp)
		}
		zones = append(zones, z)
	}
	return zones, nil
}

// Fetch 一次性擷取（自動登入），對應 Python fetch()。
func (s *Scraper) Fetch() (*Summary, []Zone, error) {
	if s.client == nil {
		if err := s.Login(); err != nil {
			return nil, nil, err
		}
	}
	summary, err := s.FetchSummary()
	if err != nil {
		return nil, nil, err
	}
	zones, err := s.FetchZones()
	if err != nil {
		return nil, nil, err
	}
	return summary, zones, nil
}

// Round2 以 IEEE half-even 對齊 Python round(x, 2)。
func Round2(v float64) float64 {
	s := strconv.FormatFloat(v, 'f', 2, 64)
	f, err := strconv.ParseFloat(s, 64)
	if err != nil {
		return math.Round(v*100) / 100
	}
	return f
}
